opus/application/static_media/js/browse.js

```diff
@@ -675,56 +674,86 @@ var o_browse = {
+            let calculatedFirstObs = (Math.floor((firstCachedObs - 1)/galleryBoundingRect.x + 0.0000001) *
+                                      galleryBoundingRect.x + 1);
             // Fot gallery view, the topBoxBoundary is the top of .gallery-contents
-            // For table view, we will set the topBoxBoundary to be the bottom of thead (account for height of thead)
+            // For table view, we will set the topBoxBoundary to be the bottom of thead
+            // (account for height of thead)
             topBoxBoundary = (topBoxBoundary ||
-                             (opus.prefs[browse]  === "gallery") ?
-                                $(`${tab} .gallery-contents`).offset().top :
-                                $(`${tab} .gallery-contents`).offset().top + $(`${tab} .op-data-table thead th`).outerHeight());
-
-            // compare the image .top + half its height in order to make sure we account for partial images
-            let topBox = $(elem).offset().top + $(elem).height()/2;
-            if (topBox >= topBoxBoundary) {
-                let obsNum = $(elem).data("obs");
-
-                // update obsNum in both infiniteScroll instances
-                // store the most top left obsNum in gallery for both infiniteScroll instances, this will be used to updated slider obsNum
-                $(`${tab} .op-gallery-view`).infiniteScroll({"obsNum": obsNum});
-                if (contentsView === ".op-data-table-view") {
-                    obsNum = Math.floor((obsNum-1)/o_browse.gallerySliderStep+0.0000001)*o_browse.gallerySliderStep+1;
+                              (opus.prefs[browse]  === "gallery") ?
+                              $(`${tab} .gallery-contents`).offset().top :
+                              $(`${tab} .gallery-contents`).offset().top + $(`${tab} .op-data-table thead th`).outerHeight());
+
+            // table: obsNum = calculatedFirstObs + number of row
+            // gallery: obsNum = calculatedFirstObs + number of row * number of obs in a row
+            let obsNumDiff = ((opus.prefs[browse] === "dataTable") ?
+                              Math.round((topBoxBoundary - firstCachedObsTop)/$(`${tab} tbody tr`).outerHeight()) :
+                              Math.round((topBoxBoundary - firstCachedObsTop)/o_browse.imageSize) * galleryBoundingRect.x);
+            let obsNum = obsNumDiff + calculatedFirstObs;
+
+            let numToDelete = ((galleryBoundingRect.x - (firstCachedObs - 1) % galleryBoundingRect.x) %
+                               galleryBoundingRect.x);
+
+            let galleryObsElem = $(`${tab} .gallery [data-obs]`);
+            let tableObsElem = $(`${tab} .op-data-table-view [data-obs]`);
+            // delete first "numToDelete" obs if row size is changed
+            if (numToDelete !== 0) {
+                for (let count = 0; count < numToDelete; count++) {
+                    o_browse.deleteCachedObservation(galleryObsElem, tableObsElem, count);
```

Comment:
I'm a little confused by this. When we delete obs 0, doesn't that make the next first obs also become obs 0? So we should be deleting obs 0 "numToDelete" times rather than deleting increasing numbers? I think this has the effect of deleting obs 0, 1, 3, 6, ... (original numbers before the delete). Actually the same comment applies to manageObservationCache. Am I confused or does this function do the wrong thing?

Reply:
The current function will do the right thing. When we pass in the galleryObsElem and tableObsElem, they are arrays of each thumbnail container and table tr. When deleting an element in an array, the element will be removed and replaced with an undefined item like the following example:
```js
let a = [6, 7, 8, 9, 10];
delete a[0];
delete a[1];
console.log(a); // [ <2 empty items>, 8, 9, 10 ]
console.log(a.length); // 5
for (let i = 0; i < a.length; i++) {
    console.log(a[i]); 
}
// the above loop will print out the following:
// undefined
// undefined
// 8
// 9
// 10
``` 
The case you mentioned above (to always delete index 0) will work if we don't pass in galleryObsElem and tableObsElem, and always re-declare both variables inside the deleteCachedObservation function like this:
```js
deleteCachedObservation: function (galleryObsElem, tableObsElem, index) {
    let galleryObsElem = $(`${tab} .gallery [data-obs]`);
    let tableObsElem = $(`${tab} .op-data-table-view [data-obs]`);
// ...
```
This will create a new array without the undefined (previous deleted obs) items.

Reply:
OK, I understand what's going on now. It works fine. Thanks.